### PR TITLE
Look at flipping clang-tidy's misc-* to enable-by-default

### PR DESCRIPTION
**Pull request by jonmeow**
I was wondering, instead of treating `misc` differently and enabling specific checks, maybe we can flip that since we actually seem okay with most of the checks?

The main check I'm enabling, with significant edits here, is `misc-no-recursion`. But maybe this is helpful to enable, even with the necessary NOLINTs, since we want to avoid recursion in the toolchain? This PR shows some example fixes in subst.cpp (which are more stylistic, since the code shouldn't actually have recursed due to its structure; I think we could remove the warning on TryResolveInst the same way). Some also just don't seem worth fixing, like those in tests files (I didn't see a way to exclude files in .clang-tidy, so instead I'm using NOLINTBEGIN). But I think we might actually want to fix inst_namer, and there's enough in convert that I didn't look closely.

Also, I made some protected -> private style fixes based on `misc-non-private-member-variables-in-classes` (this is also how I noticed `class Real` versus `struct Real`). With node_stack, it looks like the `protected` wasn't even used. [Per style](https://google.github.io/styleguide/cppguide.html#Access_Control), data members should be private outside tests. But since we can't trivially exclude `protected` members in tests, I'm turning it off -- I don't view it as offering enough benefit on the whole.

migrate_cpp issues are preexisting (I believe we just aren't monitoring it), but changes there make `bazel build --config=clang-tidy -k //...` work cleanly.

### CODE REVIEW

**Comment by jonmeow** [comment]
@zygoloid Particularly interested in your recursion thoughts here.

File: common/command_line.h [hunk 1]
```diff
@@ -307,10 +307,15 @@ class ArgBuilder {
...
-  friend CommandBuilder;
+  friend class CommandBuilder;
   // `arg` must not be null.
   explicit ArgBuilder(Arg* arg);
 
+  auto arg() -> Arg* { return arg_; }
+
+ private:
+  // TODO: Style: this should be private with a mutator.
+  // NOLINTNEXTLINE(misc-non-private-member-variables-in-classes)
```
Comment (zygoloid) [inline common/command_line.h]:
```suggestion
```

**Review by zygoloid — APPROVED** [review]

**Review by chandlerc — APPROVED** [review]
Just FYI, I like the direction, and the comments on the no-recursion bits seem pretty helpful.

**Comment by jonmeow** [comment]
Note, per discussion, I'll be looking more at misc-non-private-member-variables-in-classes but in a separate PR.

**Pull request merged**